## Bengaluru House Prediction Through Regression Models

Dataset is downloaded from here: https://www.kaggle.com/amitabhajoy/bengaluru-house-price-data

In [1]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
%matplotlib inline
import matplotlib 
matplotlib.rcParams["figure.figsize"] = (20,10)

### Loading data and analyzing the features

In [2]:
df1 = pd.read_csv('bengaluru_house_prices.csv')
df1.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00

In [3]:
df1.shape

(13320, 9)

In [4]:
df1.columns

Index(['area_type', 'availability', 'location', 'size', 'society',
       'total_sqft', 'bath', 'balcony', 'price'],
      dtype='object')

In [5]:
df1.isnull().sum()

area_type          0
availability       0
location           1
size              16
society         5502
total_sqft         0
bath              73
balcony          609
price              0
dtype: int64

In [6]:
df1.corr()

bath   balcony     price
bath     1.000000  0.204201  0.456345
balcony  0.204201  1.000000  0.120355
price    0.456345  0.120355  1.000000

In [7]:
df1['area_type'].value_counts()

Super built-up  Area    8790
Built-up  Area          2418
Plot  Area              2025
Carpet  Area              87
Name: area_type, dtype: int64

In [8]:
df1['availability'].value_counts()

Ready To Move    10581
18-Dec             307
18-May             295
18-Apr             271
18-Aug             200
                 ...  
16-Nov               1
14-Nov               1
15-Dec               1
14-Jul               1
16-Jul               1
Name: availability, Length: 81, dtype: int64

Dropping few columns which seems to be not relevant for predicting price.

    1.Area_type: More than half values are 'Super built-up' and the correlation with price seems to be very low.
    
    2.Society: Half the values are NaN and not feasible and productive to handle in NaN in this case.
    
    3.Balcony: Correlation with price is very low.
    
    4.Availability: 80% data for this column has value 'Ready to move'. Can lead to imbalanced data issues.

In [9]:
df2 = df1.drop(['area_type','society','balcony','availability'],axis='columns')

In [10]:
df2.shape

(13320, 5)

### Handling NaN values

In [11]:
df2.isnull().sum()

location       1
size          16
total_sqft     0
bath          73
price          0
dtype: int64

Data having NaN values are very low, so can be easily discarded.

In [12]:
df3 = df2.dropna()
df3.isnull().sum()

location      0
size          0
total_sqft    0
bath          0
price         0
dtype: int64

In [13]:
df3.shape

(13246, 5)

### Feature Engineering

In [14]:
df3.head()

location       size total_sqft  bath   price
0  Electronic City Phase II      2 BHK       1056   2.0   39.07
1          Chikka Tirupathi  4 Bedroom       2600   5.0  120.00
2               Uttarahalli      3 BHK       1440   2.0   62.00
3        Lingadheeranahalli      3 BHK       1521   3.0   95.00
4                  Kothanur      2 BHK       1200   2.0   51.00

In [15]:
df3['size'].value_counts()

2 BHK         5198
3 BHK         4286
4 Bedroom      818
4 BHK          577
3 Bedroom      546
1 BHK          531
2 Bedroom      329
5 Bedroom      296
6 Bedroom      191
1 Bedroom      105
8 Bedroom       84
7 Bedroom       83
5 BHK           57
9 Bedroom       46
6 BHK           30
7 BHK           17
1 RK            13
10 Bedroom      12
9 BHK            8
8 BHK            5
11 Bedroom       2
11 BHK           2
10 BHK           2
18 Bedroom       1
16 BHK           1
13 BHK           1
27 BHK           1
12 Bedroom       1
19 BHK           1
14 BHK           1
43 Bedroom       1
Name: size, dtype: int64

Here at some points BHK is used while at others Bedroom. Need to make it uniform 

In [16]:
df3['bhk'] = df3['size'].apply(lambda x: int(x.split(' ')[0]))

/home/navjot/.local/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [17]:
df3['bhk'].unique()

array([ 2,  4,  3,  6,  1,  8,  7,  5, 11,  9, 27, 10, 19, 16, 43, 14, 12,
       13, 18])

In [31]:
df3.drop(['size'], axis=1, inplace=True)

/home/navjot/.local/lib/python3.6/site-packages/pandas/core/frame.py:4174: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [32]:
df3['total_sqft'].unique()

array(['1056', '2600', '1440', ..., '1133 - 1384', '774', '4689'],
      dtype=object)

In [33]:
def isFloat(x):
    try:
        float(x)
    except:
        return False
    return True

In [34]:
df3[~df3['total_sqft'].apply(isFloat)].head(10)

location      total_sqft  bath    price  bhk
30            Yelahanka     2100 - 2850   4.0  186.000    4
122              Hebbal     3067 - 8156   4.0  477.000    4
137  8th Phase JP Nagar     1042 - 1105   2.0   54.005    2
165            Sarjapur     1145 - 1340   2.0   43.490    2
188            KR Puram     1015 - 1540   2.0   56.800    2
410             Kengeri  34.46Sq. Meter   1.0   18.500    1
549         Hennur Road     1195 - 1440   2.0   63.770    2
648             Arekere       4125Perch   9.0  265.000    9
661           Yelahanka     1120 - 1145   2.0   48.130    2
672        Bettahalsoor     3090 - 5002   4.0  445.000    4

For these(1133 - 1384, 4125Perch, 34.46Sq. Mete) type of values in total_sqft, we need to do some modifications.

In [35]:
def convert_range_to_mean(x):
    tokens = x.split('-')
    if len(tokens) == 2:
        return (float(tokens[0])+float(tokens[1]))/2
    try:
        return float(x)
    except:
        return None

In [36]:
convert_range_to_mean('5653')

5653.0

In [37]:
convert_range_to_mean('1543-5643')

3593.0

In [38]:
df4 = df3.copy()
df4['total_sqft'] = df4['total_sqft'].apply(convert_range_to_mean)
df4.head()

location  total_sqft  bath   price  bhk
0  Electronic City Phase II      1056.0   2.0   39.07    2
1          Chikka Tirupathi      2600.0   5.0  120.00    4
2               Uttarahalli      1440.0   2.0   62.00    3
3        Lingadheeranahalli      1521.0   3.0   95.00    3
4                  Kothanur      1200.0   2.0   51.00    2

In [39]:
df4['total_sqft'].unique()

array([1056. , 2600. , 1440. , ..., 1258.5,  774. , 4689. ])

In [40]:
df4[~df4['total_sqft'].apply(isFloat)].head(10)

Empty DataFrame
Columns: [location, total_sqft, bath, price, bhk]
Index: []

In [41]:
df3.shape

(13246, 5)

In [42]:
df4.loc[410]

location      Kengeri
total_sqft        NaN
bath                1
price            18.5
bhk                 1
Name: 410, dtype: object

In [43]:
df4.total_sqft.isnull().sum()

46

In [44]:
df4 = df4[df4.total_sqft.notnull()]
df4.head()

location  total_sqft  bath   price  bhk
0  Electronic City Phase II      1056.0   2.0   39.07    2
1          Chikka Tirupathi      2600.0   5.0  120.00    4
2               Uttarahalli      1440.0   2.0   62.00    3
3        Lingadheeranahalli      1521.0   3.0   95.00    3
4                  Kothanur      1200.0   2.0   51.00    2